contracts/Dispenser.sol

```diff
@@ -1,10 +1,90 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.18;
+pragma solidity ^0.8.23;
 
 import "./interfaces/IErrorsTokenomics.sol";
+import "./interfaces/IToken.sol";
 import "./interfaces/ITokenomics.sol";
 import "./interfaces/ITreasury.sol";
 
+interface IVoteWeighting {
+    function checkpointNominee(address nominee, uint256 chainId) external;
+    function nomineeRelativeWeight(address nominee, uint256 chainId, uint256 time) external returns (uint256, uint256);
+}
+
+interface ITokenomicsInfo {
+    // Structure for epoch point with tokenomics-related statistics during each epoch
+    // The size of the struct is 96 * 2 + 64 + 32 * 2 + 8 * 2 = 256 + 80 (2 slots)
+    struct EpochPoint {
+        // Total amount of ETH donations accrued by the protocol during one epoch
+        // Even if the ETH inflation rate is 5% per year, it would take 130+ years to reach 2^96 - 1 of ETH total supply
+        uint96 totalDonationsETH;
+        // Amount of OLAS intended to fund top-ups for the epoch based on the inflation schedule
+        // After 10 years, the OLAS inflation rate is 2% per year. It would take 220+ years to reach 2^96 - 1
+        uint96 totalTopUpsOLAS;
+        // Inverse of the discount factor
+        // IDF is bound by a factor of 18, since (2^64 - 1) / 10^18 > 18
+        // IDF uses a multiplier of 10^18 by default, since it is a rational number and must be accounted for divisions
+        // The IDF depends on the epsilonRate value, idf = 1 + epsilonRate, and epsilonRate is bound by 17 with 18 decimals
+        uint64 idf;
+        // Number of new owners
+        // Each unit has at most one owner, so this number cannot be practically bigger than numNewUnits
+        uint32 numNewOwners;
+        // Epoch end timestamp
+        // 2^32 - 1 gives 136+ years counted in seconds starting from the year 1970, which is safe until the year of 2106
+        uint32 endTime;
+        // Parameters for rewards and top-ups (in percentage)
+        // Each of these numbers cannot be practically bigger than 100 as they sum up to 100%
+        // treasuryFraction + rewardComponentFraction + rewardAgentFraction = 100%
+        // Treasury fraction
+        uint8 rewardTreasuryFraction;
+        // maxBondFraction + topUpComponentFraction + topUpAgentFraction <= 100%
```

Comment:
I know that we imported this as is now on tokenomics, but maybe we should update the comment saying that maxBondFraction + topUpComponentFraction + topUpAgentFraction <= 100%-serviceStakingFraction